module/builder/collection/builder.go

```diff
@@ -195,13 +195,25 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 		minRefID := refChainFinalizedID
 
 		var transactions []*flow.TransactionBody
+		var totalBytesSize uint
+		var totalGas uint64
 		for _, tx := range b.transactions.All() {
 
 			// if we have reached maximum number of transactions, stop
 			if uint(len(transactions)) >= b.config.MaxCollectionSize {
 				break
 			}
 
+			// skip the transaction (keep for the next collection)
+			if totalBytesSize >= b.config.MaxCollectionByteSize {
+				continue
+			}
+
+			// skip the transaction (keep for the next collection)
+			if totalGas >= b.config.MaxCollectionTotalGas {
+				continue
+			}
```

Comment:
@jordanschalm before I spend more time can you check if this is a good approach (skipping instead of breaking)

Reply:
We would want to `break`. If either of these conditions evaluate to true, we can't add any more transactions to the collection so we don't need to iterate any more. The other transactions will remain in the mempool for later.

Another point is that this logic will stop building a collection _after_ we exceed one of these conditions, so if we hit one of these conditions we will end up with a collection which exceeds the maximum and is therefore invalid. The approach I would take here would be more like:
```go
if totalByteSize + tx.ByteSize() > b.config.MaxCollectionByteSize { 
    break
} 
```

This would leave a bit of room for optimization, since we could leave a bit or byte/gas allocation unused by breaking early, but it's probably an OK place to start, especially if the per-collection limits are significantly larger than the per-transaction limits.